### PR TITLE
feat: Validate Signing/Verification proof purpose in DID Document 

### PR DESCRIPTION
**Pull request by rolsonquadras**
closes of #295 

Signed-off-by: Rolson Quadras <rolson.quadras@securekey.com>

### CODE REVIEW

File: pkg/doc/vc/crypto/crypto.go [hunk 1]
```diff
@@ -205,7 +224,8 @@ func (c *Crypto) SignCredential(dataProfile *vcprofile.DataProfile, vc *verifiab
 	}
 
 	signingCtx, err := c.getLinkedDataProofContext(dataProfile.DID, dataProfile.DIDKeyType,
-		dataProfile.DIDPrivateKey, dataProfile.Creator, signatureType, dataProfile.SignatureRepresentation, signOpts)
+		dataProfile.DIDPrivateKey, dataProfile.Creator, signatureType, AssertionMethod,
```
Comment (rolsonquadras) [inline pkg/doc/vc/crypto/crypto.go:227]: default assertionMethod for VC.

File: pkg/doc/vc/crypto/crypto.go [hunk 2]
```diff
@@ -234,7 +254,7 @@ func (c *Crypto) SignPresentation(profile *vcprofile.HolderProfile, vp *verifiab
 	}
 
 	signingCtx, err := c.getLinkedDataProofContext(profile.DID, profile.DIDKeyType, profile.DIDPrivateKey,
-		profile.Creator, signatureType, profile.SignatureRepresentation, signOpts)
+		profile.Creator, signatureType, Authentication, profile.SignatureRepresentation, signOpts)
```
Comment (rolsonquadras) [inline pkg/doc/vc/crypto/crypto.go:257]: default authentication for VP.

File: pkg/doc/vc/crypto/crypto.go [hunk 3]
```diff
@@ -315,6 +344,63 @@ func (c *Crypto) getSigner(did, didKeyType, didPrivateKey, creator string, opts
 	}
 }
 
+func validateProofPurpose(proofPurpose, method string, vdri vdriapi.Registry) error {
+	// TODO only trustbloc sidetree supports creating different blocks throuh uniregistrar.
+	if !strings.Contains(method, "did:trustbloc") {
```
Comment (sudeshrshetty) [inline pkg/doc/vc/crypto/crypto.go:349]: How do you know if this DID is from uniregistrar? someone might have passed private key and creator.

File: pkg/restapi/vc/operation/operations_test.go [hunk 4]
```diff
@@ -1728,11 +1728,23 @@ func TestIssueCredential(t *testing.T) {
 	kh, err := keyset.NewHandle(ecdhes.ECDHES256KWAES256GCMKeyTemplate())
 	require.NoError(t, err)
 
+	pubKey, _, err := ed25519.GenerateKey(rand.Reader)
+	require.NoError(t, err)
+
+	closeableKMS := &kmsmock.CloseableKMS{CreateSigningKeyValue: string(pubKey)}
```
Comment (fqutishat) [inline pkg/restapi/vc/operation/operations_test.go:1734]: no need

File: pkg/restapi/vc/operation/operations_test.go [hunk 5]
```diff
@@ -1728,11 +1728,23 @@ func TestIssueCredential(t *testing.T) {
...
+
+	_, signingKey, err := closeableKMS.CreateKeySet()
```
Comment (fqutishat) [inline pkg/restapi/vc/operation/operations_test.go:1736]: no need

File: pkg/restapi/vc/operation/operations_test.go [hunk 6]
```diff
@@ -2511,6 +2525,16 @@ func TestComposeAndIssueCredential(t *testing.T) {
 	})
 
 	t.Run("compose and issue credential - invalid proof format option", func(t *testing.T) {
+		pubKey, _, err := ed25519.GenerateKey(rand.Reader)
+		require.NoError(t, err)
+		closeableKMS := &kmsmock.CloseableKMS{CreateSigningKeyValue: string(pubKey)}
+
+		_, signingKey, err := closeableKMS.CreateKeySet()
```
Comment (fqutishat) [inline pkg/restapi/vc/operation/operations_test.go:2532]: no need

File: pkg/restapi/vc/operation/operations_test.go [hunk 7]
```diff
@@ -2526,7 +2550,7 @@ func TestComposeAndIssueCredential(t *testing.T) {
 			KeyManager:         &kms.KeyManager{CreateKeyValue: kh},
 			VDRI: &vdrimock.MockVDRIRegistry{
 				ResolveFunc: func(didID string, opts ...vdri.ResolveOpts) (*did.Doc, error) {
-					return createDefaultDID(), nil
+					return createDIDDocWithKeyID(didID, key1ID, base58.Decode(signingKey)), nil
```
Comment (fqutishat) [inline pkg/restapi/vc/operation/operations_test.go:2553]: use pubKey

File: pkg/doc/vc/crypto/crypto.go (hunk 3 above)
Comment (rolsonquadras) [reply pkg/doc/vc/crypto/crypto.go:349]: Well, may be I need to update comment. I meant test only for trustbloc DIDs.

File: pkg/doc/vc/crypto/crypto.go [hunk 8]
```diff
@@ -315,6 +344,63 @@ func (c *Crypto) getSigner(did, didKeyType, didPrivateKey, creator string, opts
 	}
 }
 
+func validateProofPurpose(proofPurpose, method string, vdri vdriapi.Registry) error {
+	// TODO only trustbloc sidetree supports creating different blocks throuh uniregistrar.
```
Comment (fqutishat) [inline pkg/doc/vc/crypto/crypto.go:348]: v1 support that too

File: pkg/restapi/vc/operation/operations.go [hunk 9]
```diff
@@ -742,6 +745,18 @@ func (o *Operation) createDIDUniRegistrar(keyType, signatureType string,
 		return "", "", "", fmt.Errorf("selected key not found %s", selectedKeyID)
 	}
 
+	if strings.Contains(identifier, "did:v1") {
```
Comment (troyronda) [inline pkg/restapi/vc/operation/operations.go:748]: why do we need special case for v1?

Comment (rolsonquadras) [reply pkg/restapi/vc/operation/operations.go:748]: To fetch the keys based on purpose - AssertionMethod (for issuer) and Authentication (for holder.)

File: pkg/doc/vc/crypto/crypto.go [hunk 10]
```diff
@@ -309,6 +338,63 @@ func (c *Crypto) getSigner(did, didKeyType, didPrivateKey, creator string, opts
 	}
 }
 
+// ValidateProofPurpose validates the proof purpose
+func ValidateProofPurpose(proofPurpose, method string, vdri vdriapi.Registry) error {
+	if strings.Contains(method, "did:sov") {
```
Comment (troyronda) [inline pkg/doc/vc/crypto/crypto.go:343]: need todo to revisit. did:sov supports authentication, for example.

File: pkg/restapi/vc/operation/operations.go (hunk 9 above)
Comment (troyronda) [reply pkg/restapi/vc/operation/operations.go:748]: but why is v1 special?

File: pkg/doc/vc/crypto/crypto.go (hunk 10 above)
Comment (rolsonquadras) [reply pkg/doc/vc/crypto/crypto.go:343]: created an issue https://github.com/trustbloc/edge-service/issues/368

**Pull request merged**